src/Strategy/Dart/PubSpec.hs

```diff
@@ -0,0 +1,178 @@
+module Strategy.Dart.PubSpec (
+  analyzePubSpecFile,
+
+  -- * for testing
+  buildGraph,
+  PubSpecContent (..),
+  PubSpecDepSource (..),
+) where
+
+import Control.Effect.Diagnostics (Diagnostics, context)
+import Data.Foldable (asum, for_)
+import Data.Map (Map, filterWithKey, member, notMember, toList)
+import Data.Map.Strict qualified as Map
+import Data.Maybe (fromMaybe, maybeToList)
+import Data.Text (Text)
+import Data.Yaml (FromJSON (parseJSON), (.:), (.:?))
+import Data.Yaml qualified as Yaml
+import DepTypes (
+  DepEnvironment (EnvDevelopment, EnvProduction),
+  DepType (GitType, PubType),
+  Dependency (..),
+  VerConstraint (CEq),
+ )
+import Effect.Logger (Logger (..), Pretty (pretty), logDebug)
+import Effect.ReadFS (Has, ReadFS, readContentsYaml)
+import Graphing (Graphing, direct, empty)
+import Path
+import Strategy.Dart.PubSpecLock (PackageName (..))
+import Types (GraphBreadth (..))
+
+data PubSpecContent = PubSpecContent
+  { pubSpecDependencies :: Maybe (Map PackageName PubSpecDepSource)
+  , pubSpecDevDependencies :: Maybe (Map PackageName PubSpecDepSource)
+  , pubSpecDependenciesOverrides :: Maybe (Map PackageName PubSpecDepSource)
+  }
+  deriving (Show, Eq, Ord)
+
+data PubSpecDepSource
+  = PubSpecDepHostedSource
+      { version :: Maybe VerConstraint
+      , hostedName :: Maybe Text
+      , hostedUrl :: Maybe Text
+      }
+  | PubSpecDepGitSource
+      { gitRef :: Maybe VerConstraint
+      , gitUrl :: Text
+      }
```

Comment:
Unfortunately adding record selectors to a sum type is unsafe. It generates something like:

```hs
hostedName :: PubSpecDepSource -> Maybe Text
hostedName PubSpecDepHostedSource{hostedName} = hostedName
hostedName _ = error "blah blah invalid field because wrong constructor"
```

..so we'll have to remove these field selectors, and use comments to say what the types are instead

Reply:
Another option is to use record types as parameters to sum types.

```haskell
data Sum
  = A TypeA
  | B TypeB

data TypeA = TypeA
  { valueA1 :: Int
  , valueA2 :: Int
  }

data TypeB = TypeB
  { valueB1 :: Int
  , valueB2 :: Int
  }
```

Reply:
Addr: https://github.com/fossas/spectrometer/pull/313/commits/83f5771f03c59ee7f69a3e6331b1e91eed6f5406#diff-0b87b8d82ece48b1d6e4f9d57fc9a41057de9bb5307f77bd761255d26d64fa68R58

Reply:
For posterity, this is part of the style guide: #325